tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/concurrent/PhaserHandlingParallelExecutor.java

```diff
@@ -0,0 +1,65 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+package org.eclipse.rdf4j.federated.evaluation.concurrent;
+
+import java.util.concurrent.Phaser;
+
+import org.eclipse.rdf4j.common.iteration.CloseableIteration;
+import org.eclipse.rdf4j.federated.structures.QueryInfo;
+import org.eclipse.rdf4j.query.BindingSet;
+import org.eclipse.rdf4j.query.QueryEvaluationException;
+
+/**
+ * A delegating {@link ParallelExecutor} which arrives and de-registers on the phaser on completion of a task.
+ * 
+ * @author Andreas Schwarte
+ *
+ */
+public class PhaserHandlingParallelExecutor implements ParallelExecutor<BindingSet> {
```

Comment:
Any chance we can make this package-private? If not, perhaps slap an `@InternalUseOnly` annotation on it.

Reply:
I have moved it to the `org.eclipse.rdf4j.federated.evaluation.join` package and made it package private.